[Lye Jia Yang] iP

Jiaaa-yang opened the pull request:
# Duke v2
> "A Duke that works" - Yours truly

**Duke v2** is a simple upgrade over the visual concept of this upstream repo, with no fancy changes so that you still get your good old _Duke_,  just better!

It focuses on:
* Ease of use
* Robust internal system
* Good recovery from errors

Follow these simple steps to try out the pre-release!
- [ ] Download this [jar file](https://github.com/Jiaaa-yang/ip/releases/tag/A-Jar)
- [ ] Double click the file
- [ ] Read the user guide to see what is supported (coming soon)
- [ ] Start tracking your tasks!

Remember, to succeed in life 🪜 you need just 3 steps:
1. Add a task
2. Plan for the task
3. Do the task

and **Duke v2** will be there with you for your whole journey!

---
Also check out the main entry class `Duke`, with a simple but effective abstraction!
```java
class Duke {
    private static final Path DEFAULT_FILE_PATH = Paths.get("data", "duke.txt");

    public static void main(String[] args) {
        Storage store = Storage.initStorage(DEFAULT_FILE_PATH);
        TaskList taskList = TaskList.initTaskList(store);

        ChatBot chatBot = new ChatBot(taskList);
        chatBot.initialise();

        Scanner scanner = new Scanner(System.in);
        while (!chatBot.hasTerminated()) {
            String input = scanner.nextLine();
            chatBot.runCommand(input);
        }
    }
}
```

Commit:
Add Gradle support

Jiaaa-yang committed:
feat: add greet, echo and exit
Added ui package to group chatbot interface implementations.

Jiaaa-yang committed:
feat: add storing and listing of user text

Jiaaa-yang committed:
refactor: move input parsing to Command class
Added method to parse user input command to separate
name and arguments.

Jiaaa-yang committed:
feat: add ability to change task done status

Jiaaa-yang committed:
feat: add differentiation for type of tasks
Changed Task to abstract class, and implemented
ToDo, Event and Deadline subclasses to represent
different tasks.

Jiaaa-yang committed:
test: add regression test with txt file

Jiaaa-yang committed:
feat: add error handling of invalid inputs

Jiaaa-yang committed:
feat: add deletion of tasks

Jiaaa-yang committed:
refactor: follow module coding standards

Update the following based on given code style/standard:
* change wildcard imports to single class imports
* remove indent for cases after switch statement
* rename boolean variables to sound like one

Jiaaa-yang committed:
Add ability to save tasks on disk

Use text file with predefined format to maintain
state of task list during and between ChatBot
operations.

Jiaaa-yang committed:
Add date parsing for deadlines

Allow users to pass in deadline in specified format,
with error checking. Also displays it in a more readable
way in task list.

Jiaaa-yang committed:
Update test for deadline dates and delete command

Jiaaa-yang committed:
Merge branch 'branch-Level-7'

Jiaaa-yang committed:
Merge branch 'branch-Level-8'

Jiaaa-yang committed:
Add data package to deal with task state

Implement Storage and TaskList class to sync operations
on task list with backing file storage.

Jiaaa-yang committed:
Refactor logic to use data package

Make use of TaskList class to better encapsulate list
of tasks maintained by ChatBot. Also use Storage class
to ensure consistency between TaskList state and
disk file data.

Jiaaa-yang committed:
Update test script to delete old data file

Fix bug with regression test where it was failing
due to data from previous tests.

Jiaaa-yang committed:
Add gradle support

Jiaaa-yang committed:
Add JUnit tests for task package

Jiaaa-yang committed:
Configure gradle for shadowJar task

Jiaaa-yang committed:
Update JavaDocs to meet module standards

Change style of JavaDocs comment to follow coding
standards laid out for the module:
* Add period for parameters description
* Change class member Javadoc to single line

Jiaaa-yang committed:
Refactor code style by coding standard

Fix the following style issues following given
coding standard:
* Add common prefix for command feedback constants
* Remove unused imports

Jiaaa-yang committed:
Add functionality to search task by keyword

Jiaaa-yang committed:
Merge branch 'branch-A-JavaDoc'

Jiaaa-yang committed:
Merge branch 'branch-A-CodingStandard'

Jiaaa-yang committed:
Merge branch 'branch-Level-9'

Jiaaa-yang committed:
Add Javadocs for overriden methods

Add comments for overriden methods in subclasses
of Command to explain what each command does
when executed.

Jiaaa-yang committed:
Merge branch 'add-gradle-support' of https://github.com/Jiaaa-yang/ip into branch-A-Gradle

Jiaaa-yang committed:
Add config file for checkstyle

Jiaaa-yang committed:
Refactor code to fix checkstyle violations

Fix all errors under main with exceptions of Javadocs
and import statement orders, which will be left to later.

Jiaaa-yang committed:
Move main Duke class to duke package

Jiaaa-yang committed:
Add initial support for GUI

Initial version of GUI with minimal styling mostly
from provided JavaFX tutorial.

Jiaaa-yang committed:
Update response logic in application

Change ChatBot to return response to application for
display instead of printing to stdout. Also renamed ui
package to chatbot as it is no longer in charge of ui.

Jiaaa-yang committed:
Fix application not closing on exit command

Jiaaa-yang committed:
Use FXML for application

Change structuring of JavaFX application to use
FXML for main layout and dialog box.

simjunheng commented on src/main/java/data/Storage.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    /**
+     * Initialise a Storage instance linked to
```
Should the word **Initialise** be changed into **Initialises**?

Jiaaa-yang replied:
Yeap that should be changed! Thanks for pointing this out and some other places as well as I have overlooked them!

simjunheng commented on src/main/java/data/Storage.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    /**
+     * Read all lines of linked filePath to
```
Same as above. Perhaps you can consider changing it into **Reads**

simjunheng commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    /**
+     * Gets the name of the Task.
```
Are descriptive header comments needed for getters/setters?

simjunheng commented on src/main/java/ui/ChatBot.java:
```diff
@@ -0,0 +1,57 @@
+            Command command = Command.parseCommand(input, this.taskList);
+            this.hasTerminated = command.execute();
+        } catch (IllegalArgumentException e) {
+            ArrayList<String> response = new ArrayList<>();
```
Should we use **responses** instead of **response** since it is representing a collection of objects?

Jiaaa-yang replied:
Thanks for pointing this out! **responses** might not be appropriate since it is a single response by my ChatBot to a command, but I do agree that I should change it to reflect its state as a collection.

simjunheng commented on src/main/java/ui/command/Command.java:
```diff
@@ -0,0 +1,128 @@
+ * An abstract class to represent the commands
+ * that can be issued to a ChatBot.
+ */
+public abstract class Command {
```
Do you think Parser is a better name for this class instead of Command? It seems that the main purpose of this class is to parse the inputs given by the users.

Jiaaa-yang replied:
Yeah the logic of parsing inputs could be placed in a `Parser` class instead to better separate my application logic.

simjunheng commented on src/main/java/ui/command/TaskListCommand.java:
```diff
@@ -0,0 +1,21 @@
+ * Command which operates on the TaskList
+ * maintained by ChatBot
+ */
+abstract class TaskListCommand extends Command {
```
In your view, do you think this class can be combined with the Command class?

Jiaaa-yang replied:
It does seem that this class seems to be redundant now. It was initially used to separate commands that required access to the task list and those that did not, but that separation has not much purpose.

simjunheng reviewed (commented):
Overall, I didn't detect any serious issues in terms of coding standards. Well done! 😄

Jiaaa-yang committed:
Add styling for application

* Add background colors for all components
* Round image for duke and user avatar
* Add border to show text better

Jiaaa-yang committed:
Allow application to exit gracefully

Update exit command to disable further inputs
for application to allow user to quit themselves,
instead of closing the application forcefully.

Jiaaa-yang committed:
Merge branch 'branch-Level-10'

s7manth commented on src/main/java/duke/chatbot/ChatBot.java:
```diff
@@ -0,0 +1,63 @@
+     * of an ArrayList of string.
+     *
+     * @param input String command by user to run.
+     * @return ArrayList of string containing response of command ran.
```
One suggestion, you could have followed one convention overall. Like you can begin all descriptions for `@params`, `@throws` and `@return` with capital letters.

s7manth commented on src/main/java/duke/chatbot/command/Command.java:
```diff
@@ -0,0 +1,129 @@
+     * @param input Command inputted by user.
+     * @param taskList TaskList maintained by ChatBot.
+     * @return Command object corresponding to input.
+     * @throws IllegalArgumentException If the command is not valid.
```
Very descriptive Java Docs! Good job!

s7manth commented on src/test/java/duke/task/DeadlineTest.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @Test
+    void getDescription_sampleDeadline_matchExpected() {
```
Very nice! Following the conventions.

s7manth reviewed (commented):
LGTM

Jiaaa-yang committed:
Fix empty message on list command

List command was showing an empty dialog
if list was empty. Change it to show a more
descriptive message.

Jiaaa-yang committed:
Add assertions for unverified assumptions

Jiaaa-yang committed:
Delete dead code

Remove EchoCommand from initial iteration,
and also functionalities related to old text-based
user interface.

Jiaaa-yang committed:
Combine mark and unmark task command

The logic provided by MarkTask and UnmarkTaskCommand
class is nearly identical.

This resulted in high code duplication in the two tasks.

Let's combine the two classes by extracting out the logic
to differentiate between the two commands, so that the
combined class can perform either one of them.

Jiaaa-yang committed:
Refactor to adhere to coding quality guidelines

The code has some unsafe practices such as no default
branches for switch statements, and general violations
affecting readability.

This can result in bugs and issue with maintainbility
in the future.

Let's refactor the code to fix these issues:
* Add default branch for switch statement to raise errors
* Follow SLAP in methods
* Update variable and method names to be more descriptive

Jiaaa-yang committed:
Merge pull request #2 from Jiaaa-yang/branch-A-CodeQuality

Refactor to improve code quality

Jiaaa-yang committed:
Merge branch 'master' into branch-A-Assertions

Jiaaa-yang committed:
Merge pull request #1 from Jiaaa-yang/branch-A-Assertions

Add assertions

Jiaaa-yang committed:
Refactor to fix checkstyle errors

Reorder import statements based on checkstyle file
and add missing JavaDocs for public members.

Jiaaa-yang committed:
Add ability to undo last command

Jiaaa-yang committed:
Refactor parsing of task index to parent class

Jiaaa-yang committed:
Fix bug for undoing delete commands

Undo of delete commands was adding deleted tasks
back to end of list. Fixed it to add deleted task
back at its original position.

Jiaaa-yang committed:
Merge branch 'branch-C-Undo'

Jiaaa-yang committed:
Update style for crediting for reused code

Change credit of reused code to fit style laid
out by module's reuse policy.

Jiaaa-yang committed:
Update Duke feedback to fix grammatical error

Jiaaa-yang committed:
Update user guide and add product screenshot

Jiaaa-yang committed:
Fix typo in user guide
